### PR TITLE
Implement imported target for FLANN

### PR DESCRIPTION
**Pull request by SunBlack**
Base of documentation and coding style is shipped FindGTest.cmake script of CMake

### CODE REVIEW

**Comment by SunBlack** [comment]
Not sure if we really need `-DFLANN_STATIC` (just added it again, because I'm not sure)

File: doc/tutorials/content/sources/vfh_recognition/FindFLANN.cmake [hunk 1]
```diff
@@ -1,21 +1,42 @@
-###############################################################################
```
Comment (SergioRAgostinho) [inline doc/tutorials/content/sources/vfh_recognition/FindFLANN.cmake]: The existence of this file baffles me. I propose to remove it and add PCL's cmake modules folder to the `CMAKE_MODULE_PATH` path of this tutorial so that it makes use of the same file. This if accepted, to be done in a separate commit.

**Comment by SergioRAgostinho** [comment]
> Not sure if we really need `-DFLANN_STATIC` (just added it again, because I'm not sure)

I don't see an explicit use in our source files. But I have the impression this might be needed by FLANN and its dependencies. Actually that's the feeling that it comes out from this [PR](https://github.com/mariusmuja/flann/pull/399/files) @jspricke submitted, specifically the addition of the LZ4_STATIC_FLAGS.

**Comment by claudiofantacci** [comment]
> Ping @claudiofantacci to monitor this one.

Yes, I think it does. By a quicklook at the PR, it creates a FLANN target with all the required info, just as the CMake file I was suggesting to use to address the problem ([StandardFindModule.cmake](https://github.com/robotology/ycm/blob/master/modules/StandardFindModule.cmake) from [YCM](https://github.com/robotology/ycm)).

File: cmake/Modules/FindFLANN.cmake [hunk 2]
```diff
@@ -1,63 +1,145 @@
...
+
 if(FLANN_FOUND)
-  message(STATUS "FLANN found (include: ${FLANN_INCLUDE_DIRS}, lib: ${FLANN_LIBRARIES})")
-  if(FLANN_USE_STATIC)
-    add_definitions(-DFLANN_STATIC)
+  if(NOT TARGET FLANN::FLANN)
+    if (FLANN_USE_STATIC)
+      add_library(FLANN::FLANN STATIC IMPORTED)
+    else()
+      add_library(FLANN::FLANN SHARED IMPORTED)
```
Comment (claudiofantacci) [inline cmake/Modules/FindFLANN.cmake]: Just one comment here: if I'm not wrong, we should add the `INTERFACE` keyword to `add_library` (the same applies to the `STATIC` case). See also the CMake [socumentation](https://cmake.org/cmake/help/v3.5/command/add_library.html?#interface-libraries) about this.

File: cmake/Modules/FindFLANN.cmake [hunk 3]
```diff
@@ -1,63 +1,145 @@
...
-    find_library(FLANN_LIBRARY_DEBUG
-                 NAMES ${FLANN_DEBUG_NAME} ${FLANN_RELEASE_NAME}
-                 HINTS "${FLANN_ROOT}" "$ENV{FLANN_ROOT}"
-                 PATHS "$ENV{PROGRAMFILES}/flann" "$ENV{PROGRAMW6432}/flann"
-                 PATH_SUFFIXES lib debug/lib)
-
-    if(NOT FLANN_LIBRARY_DEBUG)
-      set(FLANN_LIBRARY_DEBUG ${FLANN_LIBRARY})
-    endif()
+find_path(FLANN_INCLUDE_DIR
```
[the remark this answers is not in the document]
Comment (SergioRAgostinho) [reply cmake/Modules/FindFLANN.cmake]: Well, as it is, this will fix bug #2802. This can be circumvented if we are able to generate full paths to the libraries from the information pkg-config provides. In case it's a couple of libs in a single folder this is trivial, but if it's a couple of libs from multiple folders it becomes more troublesome.

The need for having absolute paths to libs comes from CMake convention. Quoting from the [CMake docs page on link_directories](https://cmake.org/cmake/help/v3.0/command/link_directories.html).
> Note that this command is rarely necessary. Library locations returned by find_package() and find_library() are absolute paths. Pass these absolute library file paths directly to the target_link_libraries() command. CMake will ensure the linker finds them.

This expectation is so prevalent that there's no target interface properties regarding the lib directories. [Here's](https://cmake.org/cmake/help/v3.0/manual/cmake-properties.7.html?highlight=target%20properties#properties-on-targets) the extensive list. This becomes a problem for our downstream projects which are not explicitly using FLANN. It forces us to add an unconditional `link_libraries(FLANN_LIBRARY_DIRS)` during PCLConfig.cmake which is a little less clean.

There's also the option of setting linker flags and passing `-L` for nix systems and whatever is the equivalent on Windows. It feels hacky to me though.

[the remark this answers is not in the document]
Comment (jspricke) [reply cmake/Modules/FindFLANN.cmake]: You are right, this is documented for CMake here: https://cmake.org/cmake/help/latest/manual/cmake-developer.7.html#a-sample-find-module

sorry for the confusion.

File: CMakeLists.txt [hunk 4]
```diff
@@ -276,7 +276,6 @@ if(NOT PCL_SHARED_LIBS OR ((WIN32 AND NOT MINGW) AND NOT PCL_BUILD_WITH_FLANN_DY
   set(FLANN_USE_STATIC ON)
 endif()
 find_package(FLANN 1.7.0 REQUIRED)
-include_directories(SYSTEM ${FLANN_INCLUDE_DIRS})
```
[the remark this answers is not in the document]
Comment (jspricke) [reply CMakeLists.txt]: resolved.

File: cmake/Modules/FindFLANN.cmake (hunk 2 above)
Comment (SunBlack) [reply cmake/Modules/FindFLANN.cmake]: i created a ticket at CMake repo for [this](https://gitlab.kitware.com/cmake/cmake/issues/18975), because i don't see when to use Interface Imported Target and when to use Imported Target.

File: doc/tutorials/content/sources/vfh_recognition/FindFLANN.cmake (hunk 1 above)
Comment (SunBlack) [reply doc/tutorials/content/sources/vfh_recognition/FindFLANN.cmake]: Was wondering too ;)

File: cmake/Modules/FindFLANN.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (SunBlack) [reply cmake/Modules/FindFLANN.cmake]: So nothing to change here and I can remove `link_libraries(FLANN_LIBRARY_DIRS)`?

[the remark this answers is not in the document]
Comment (jspricke) [reply cmake/Modules/FindFLANN.cmake]: Yes :).

File: cmake/Modules/FindFLANN.cmake (hunk 2 above)
Comment (SunBlack) [reply cmake/Modules/FindFLANN.cmake]: I got an answer: Imported target without `INTERFACE` keyword is correct.

Comment (claudiofantacci) [reply cmake/Modules/FindFLANN.cmake]: Ok, thanks!

File: doc/tutorials/content/sources/vfh_recognition/FindFLANN.cmake (hunk 1 above)
Comment (SergioRAgostinho) [reply doc/tutorials/content/sources/vfh_recognition/FindFLANN.cmake]: Hey. You marked this as resolved despite not being resolved yet, the duplicate file is still here. Just wanted to raise attention to this.

Comment (SunBlack) [reply doc/tutorials/content/sources/vfh_recognition/FindFLANN.cmake]: Ah believed separate commit = separate PR. My fault ;).

**Comment by SunBlack** [comment]
> Let's discuss squash strategy. I would propose to squash the first three commits and leave the last one separate.

I moved every changed in vfh_recognition/CMakeLists.txt to second commit, too ;).

File: PCLConfig.cmake.in [hunk 5]
```diff
@@ -645,7 +645,7 @@ endif()
 pcl_remove_duplicate_libraries(PCL_COMPONENTS PCL_LIBRARIES)
 
 # Add 3rd party libraries, as user code might include our .HPP implementations
-list(APPEND PCL_LIBRARIES ${BOOST_LIBRARIES} ${QHULL_LIBRARIES} ${OPENNI_LIBRARIES} ${OPENNI2_LIBRARIES} ${ENSENSO_LIBRARIES} ${davidSDK_LIBRARIES} ${DSSDK_LIBRARIES} ${RSSDK_LIBRARIES} ${FLANN_LIBRARIES} ${VTK_LIBRARIES})
+list(APPEND PCL_LIBRARIES ${BOOST_LIBRARIES} ${QHULL_LIBRARIES} ${OPENNI_LIBRARIES} ${OPENNI2_LIBRARIES} ${ENSENSO_LIBRARIES} ${davidSDK_LIBRARIES} ${DSSDK_LIBRARIES} ${RSSDK_LIBRARIES} FLANN::FLANN ${VTK_LIBRARIES})
```
Comment (taketwo) [inline PCLConfig.cmake.in]: We will have a problem with this line if PCL is compiled without components that need Flann. In this case the `find_flann` macro is never called, Flann is never searched for, and this target is undefined. Previously this worked because `${FLANN_LIBRARIES}` would be undefined and thus treated as an empty string.

A possible solution is to undo change in this line and set `FLANN_LIBRARIES` to `FLANN::FLANN` inside `find_flann` macro.

Comment (SunBlack) [reply PCLConfig.cmake.in]: Used another workaround for this now ;)
